File: samples/java/DeliveryDispatch/Server/CustomerGateway/src/main/java/systems/zlink/samples/deliverydispatch/server/customergateway/CustomerGatewayApplication.java
package systems.zlink.samples.deliverydispatch.server.customergateway;

import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.boot.context.properties.EnableConfigurationProperties;
import org.springframework.context.annotation.Bean;

import systems.zlink.contracts.core.RoutingId;
import systems.zlink.framework.configuration.ZLinkMeshNodeBuilder;
import systems.zlink.framework.configuration.ZLinkMessageFlowLogMode;
import systems.zlink.framework.locations.redis.ZLinkRedisLocationStore;
import systems.zlink.framework.monitoring.ZLinkRouteMeshRuntime;
import systems.zlink.framework.spring.EnableZLinkFramework;
import systems.zlink.framework.spring.ZLinkFrameworkConfigurer;
import systems.zlink.samples.deliverydispatch.server.configuration.DeliveryDispatchReadinessReporter;
import systems.zlink.samples.deliverydispatch.server.configuration.SampleApplication;
import systems.zlink.samples.deliverydispatch.server.configuration.SampleLocationStore;
import systems.zlink.samples.deliverydispatch.server.configuration.SampleNames;
import systems.zlink.samples.deliverydispatch.server.configuration.SampleTopology;
import systems.zlink.samples.deliverydispatch.server.customergateway.sessions.CustomerSession;
import systems.zlink.samples.deliverydispatch.server.customergateway.spots.CustomerEntrySpot;

@EnableZLinkFramework
@EnableConfigurationProperties(SampleTopology.class)
@SpringBootApplication(
        proxyBeanMethods = false,
        scanBasePackageClasses = CustomerGatewayApplication.class)
public final class CustomerGatewayApplication {
    private CustomerGatewayApplication() {}

    public static AutoCloseable run(String configPath) {
        return SampleApplication.start(CustomerGatewayApplication.class, configPath)::close;
    }

    @Bean
    ZLinkFrameworkConfigurer customerGatewayFramework(SampleTopology topology) {
        return options -> {
            options.addHandlersFromPackageOf(CustomerGatewayApplication.class);
            options.configureDispatch().messageFlow(ZLinkMessageFlowLogMode.NORMAL);
            ZLinkMeshNodeBuilder node = options.addRouteMesh(SampleNames.CustomerSpotDiscovery);
            node.listen(topology.customerSpotRouterEndpoint())
                    .setRoutingId(RoutingId.from(SampleNames.CustomerGatewayNode));
            node.objects()
                    .server()
                    .addEntrySpot(CustomerEntrySpot.class)
                    .addActorFactory(
                            SampleNames.CustomerActorType,
                            CustomerActor.class,
                            CustomerActorFactory.class,
                            factory -> factory.disableRelocation());
            options.addStreamNode(SampleNames.CustomerStreamNode)
                    .bind(topology.customerStreamEndpoint())
                    .enableActorDispatch()
                    .registerSession(CustomerSession.class);
        };
    }

    @Bean(destroyMethod = "close")
    ZLinkRedisLocationStore locationStore(SampleTopology topology) {
        return SampleLocationStore.create(topology);
    }

    @Bean
    DeliveryDispatchReadinessReporter readinessReporter(ZLinkRouteMeshRuntime routeMeshRuntime) {
        return new DeliveryDispatchReadinessReporter(
                routeMeshRuntime,
                SampleNames.CustomerGatewayNode,
                SampleNames.CustomerSpotDiscovery);
    }

    @Bean
    CustomerActorDirectory customerActorDirectory() {
        return new CustomerActorDirectory();
    }
}

File: samples/kotlin/DeliveryDispatch/Server/CustomerGateway/src/main/kotlin/systems/zlink/samples/kotlin/deliverydispatch/server/customergateway/CustomerGatewayApplication.kt
package systems.zlink.samples.kotlin.deliverydispatch.server.customergateway

import kotlinx.coroutines.Dispatchers
import org.springframework.boot.WebApplicationType
import org.springframework.boot.autoconfigure.SpringBootApplication
import org.springframework.boot.builder.SpringApplicationBuilder
import org.springframework.context.annotation.Bean
import systems.zlink.framework.configuration.ZLinkMessageFlowLogMode
import systems.zlink.framework.kotlin.useCoroutineHandlers
import systems.zlink.framework.locations.redis.ZLinkRedisLocationStore
import systems.zlink.framework.monitoring.ZLinkRouteMeshRuntime
import systems.zlink.framework.spring.EnableZLinkFramework
import systems.zlink.framework.spring.ZLinkFrameworkConfigurer
import systems.zlink.samples.kotlin.deliverydispatch.server.configuration.DeliveryDispatchReadinessReporter
import systems.zlink.samples.kotlin.deliverydispatch.server.configuration.SampleLocationStore
import systems.zlink.samples.kotlin.deliverydispatch.server.configuration.SampleNames
import systems.zlink.samples.kotlin.deliverydispatch.server.configuration.SampleTopology
import systems.zlink.samples.kotlin.deliverydispatch.server.customergateway.sessions.CustomerSession
import systems.zlink.samples.kotlin.deliverydispatch.server.customergateway.spots.CustomerEntrySpot

@EnableZLinkFramework
@SpringBootApplication(
    proxyBeanMethods = false,
    scanBasePackageClasses = [CustomerGatewayApplication::class],
)
class CustomerGatewayApplication {
    @Bean
    fun customerGatewayFramework(): ZLinkFrameworkConfigurer = ZLinkFrameworkConfigurer { options ->
        options.useCoroutineHandlers(Dispatchers.Default)
        options.addHandlersFromPackageOf(CustomerGatewayApplication::class.java)
        options.configureDispatch().messageFlow(ZLinkMessageFlowLogMode.NORMAL)

        val node = options.addRouteMesh(SampleNames.CustomerSpotMesh)
        node
            .listen(SampleTopology.CustomerSpotRouterEndpoint)
            .setRoutingIdPrefix("delivery-customer")
        node.objects().server().addEntrySpot(CustomerEntrySpot::class.java).addActorFactory(
            SampleNames.CustomerActorType,
            CustomerActor::class.java,
            CustomerActorFactory::class.java,
        ) { factory ->
            factory.disableRelocation()
        }
        options
            .addStreamNode(SampleNames.CustomerStreamNode)
            .bind(SampleTopology.CustomerStreamEndpoint)
            .enableActorDispatch()
            .registerSession(CustomerSession::class.java)
    }

    @Bean fun locationStore(): ZLinkRedisLocationStore = SampleLocationStore.create()

    @Bean(destroyMethod = "close")
    fun readinessReporter(meshes: ZLinkRouteMeshRuntime): DeliveryDispatchReadinessReporter =
        DeliveryDispatchReadinessReporter.route(
            SampleNames.CustomerGatewayNode,
            SampleNames.CustomerSpotMesh,
            meshes,
        )

    @Bean fun customerActorDirectory(): CustomerActorDirectory = CustomerActorDirectory()

    companion object {
        fun run(args: Array<String> = emptyArray()): AutoCloseable {
            val builder =
                SpringApplicationBuilder(CustomerGatewayApplication::class.java)
                    .web(WebApplicationType.NONE)
            builder.application().setKeepAlive(true)
            val context = builder.run(*args)
            return AutoCloseable { context.close() }
        }
    }
}
